heka/io_modules/decoders/gelf.lua

```diff
@@ -0,0 +1,120 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+--[[
+# Heka Gelf Decoder Module
+Takes a gelf message, which should be a UTF-8 string, and adds it to the Heka
+message Payload header.
+
+## Decoder Configuration Table
+- none
+
+## Functions
+
+### decode
+
+Decode and inject the resulting message
+
+*Arguments*
+- data (string) - Data to write to the msg.Payload
+- default_headers (optional table) - Heka message table containing the default
+  header values to use, if they are not populated by the decoder. If 'Fields'
+  is specified it should be in the hashed based format see:
+  http://mozilla-services.github.io/lua_sandbox/heka/message.html
+
+*Return*
+- nil - throws an error on inject_message failure.
+
+--]]
+
+-- Imports
+
+local module_name    = ...
+local module_cfg     = require "string".gsub(module_name, "%.", "_")
+local cjson          = require "cjson"
+local inject_message = inject_message
+local ipairs         = ipairs
+local pairs          = pairs
+local type           = type
+local table          = require "table"
+local string         = require "string"
+local io = require "io"
+local read_config    = read_config
+
+local cfg = read_config(module_cfg) or {}
+
+
+local map = {}
+local function table_length(t)
+   local c = 0
+   for k,v in pairs(t) do
+      c = c+1
+   end
+   return c
+end
+
+local M = {}
+setfenv(1, M) -- Remove external access to contain everything in the module
+
+local function build_and_send_message(data, dh)
+   local msg_type = read_config("type")
+   local json = cjson.decode(data)
+   local msg = {
```

Comment:
msg should be pulled outside of this function since there in no need to GC and reconstruct the skeleton each time